tools/dotnet-linker/Steps/ManagedRegistrarStep.cs

```diff
@@ -459,12 +461,57 @@ public void EmitCallToExportedMethod (MethodDefinition method, MethodDefinition
 					// We're throwing an exception, so there's no need for any more code.
 					skipEverythingAfter = il.Body.Instructions.Last ();
 				} else {
-					il.Emit (OpCodes.Ldarg_0);
+					// Whenever there's an NSObject constructor that we call from a registrar callback, we need to create
+					// a separate constructor that will first set the `handle` and `flags` values of the NSObject before
+					// calling the original constructor. Here's an example of the code we generate:
+					//
+					// // The original constructor:
+					// public .ctor (T0 p0, T1 p1, ...) { /* ... */ }
+					//
+					// // The generated constructor with pre-initialization: 
+					// public .ctor (NSObjectInitializationData initData, T0 p0, T1 p1, ...) {
+					//     this.handle = initData.Handle;
+					//     this.flags = initData.Flags;
+					//     this..ctor (p0, p1, ...);
+					// }
```

Comment:
Just from a performance perspective, I think it would make sense to add the additional data at the end:

```suggestion
					// public .ctor (T0 p0, T1 p1, ..., NSObjectInitializationData initData) {
					//     this.handle = initData.Handle;
					//     this.flags = initData.Flags;
					//     this..ctor (p0, p1, ...);
					// }
```

because that way the forwarded arguments wouldn't have to be shuffled around in the registers to call the final constructor.

In other words: T0 will be in x0 (for arm64) during the entire method call, and not have to be moved from x2 to x0 when calling the final constructor.